### PR TITLE
`Programming exercises`: Show test case details consistently for tutors on all related pages

### PR DESCRIPTION
**Pull request by Strohgelaender**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.

### Motivation and Context
See #5959

When an editor or instructor is working with automatic feedback on Artemis, there currently is not consistent behavior as to whether or not the test case names are shown to the user. For example, when viewing a student's score via the Scores-Page, the test case names are shown, but when using the submissions-page instead, no more names are present.

### Description
This PR aims to create a more unified behavior. Instead of allowing every component to pass a boolean from the outside, now the feedback component itself determines if test case names should be shown.

Test case names now get consistently shown if either: 
- The programming exercise has the option "show test case names to students" active
- The user is a least an Editor in the course

Test all places where a PE result is shown:
- The result view on the top right of the exercise
- Clicking on a task name in the problem statement
- Clicking on "show more results"
- Clicking on the task icons on the top of the page
- While using the online code editor 

For editors, also check the following places (names should always be show)
- Scores table
- Submissions Table
- Individual Submissions Table (Click on the number of submissions on the scores table)
- Template / Solution Result
- Template / Solution Submission (Click on "Show Submissions" under the result)
- Assessment Dashboard (for exercises with manual correction)
- Assessment Code Editor (for exercises with manual correction)

### Steps for Testing
Prerequisites:
- 2 Programming Exercises ("Show test case names to students off and on")
- 1 Tutor
- 1 Student

Check that the student only sees test case names if the option is set on.

#### Exam Mode Testing
Similar to above, test that students cannot set test case names while solving an exam exercise or when reviewing the exam results.

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Exam Mode Test
- [x] Test

### Test Coverage
all changed covered

### CODE REVIEW

**Review by Vinicius-Agreste — DISMISSED** [review]
Under this configuration (Exercise Editor + "show test case names to students" inactive) only the following places do not show the test names:
- The result view on the top right of the exercise
- Clicking on a task name in the problem statement
- Individual Submissions Table (Click on the number of submissions on the scores table)
Student perspective as well as other configurations look good. 
Waiting for this change in order to test exam.

**Review by simon-hng — APPROVED** [review]
Code

**Review by PhilipWerz — DISMISSED** [review]
Everything works as expected for normal exercise and for exam exercise. Does not show testcase names anymore if field is active. Tested on ts0.

**Review by JohannesStoehr — DISMISSED** [review]
Code looks good and unifies the behavior.
One could argue that a tutor that participates in the exercise should not see the test names in the participation view for two reasons:
- Wenn participating themselves tutors should see the same as student
- In case the tutor streams the participation during a tutor session, students might be confused about the test names, which they might not be seeing

**Comment by Strohgelaender** [comment]
I was thinking about this as well (and that's why I initially coded that only editors should see the names). However, I think that during the assessment of a programming exercise, seeing the test case names is important to have more insight into what causes tests to fail and how to grade this.

Distinguishing between a tutor grading an exercise and a tutor participating in an exercise would re-introduce the outside parameter, which my goal was to remove in this PR.

What are other opinions on this?

**Review by Vinicius-Agreste — DISMISSED** [review]
Tested on ts1. Worked as expected. Working in Exam mode as well.

**Review by simon-hng — DISMISSED** [review]

File: src/main/webapp/app/exercises/shared/feedback/feedback.component.ts [hunk 1]
```diff
@@ -154,6 +154,9 @@ export class FeedbackComponent implements OnInit {
         if (!this.exerciseType && isProgrammingExerciseParticipation(this.result?.participation)) {
             this.exerciseType = ExerciseType.PROGRAMMING;
         }
+
+        this.showTestDetails =
+            this.exercise?.isAtLeastTutor || (this.exerciseType === ExerciseType.PROGRAMMING && (this.exercise as ProgrammingExercise)?.showTestNamesToStudents) || false;
```
Comment (krusche) [inline src/main/webapp/app/exercises/shared/feedback/feedback.component.ts:159]: The PR description refers to editors, here you have tutors. What is the correct role to see the test names?

**Review by krusche — CHANGES_REQUESTED** [review]
Ideally the server would not even send the test case names in case they should not be visible, otherwise students could look it up in the REST response.

Could we set the testcase names to null on the server before sending it to the client? 
Then, the client could simply display them, in case they are available and all logic is on the server, making it easier for the newly developed Android and iOS apps to stay consistent

File: src/main/webapp/app/exercises/shared/feedback/feedback.component.ts (hunk 1 above)
Comment (Strohgelaender) [reply src/main/webapp/app/exercises/shared/feedback/feedback.component.ts:159]: See the discussion with @JohannesStoehr above. I initially only changed it to editors, but some pages like the assessment code editor were already showing test case names to tutors. So I decided to show test case names also to tutors consistently.

**Review by tobias-lippert — APPROVED** [review]
Code LGTM.

**Review by JohannesStoehr — APPROVED** [review]
Newest code changes look good to me

File: src/main/java/de/tum/in/www1/artemis/domain/Result.java [hunk 2]
```diff
@@ -470,19 +470,37 @@ public void filterSensitiveInformation() {
...
+
+        if (exercise instanceof ProgrammingExercise programmingExercise) {
+            var testCaseFeedback = feedbacks.stream().filter(Feedback::isTestFeedback).toList();
+            if (!programmingExercise.getShowTestNamesToStudents()) {
+                testCaseFeedback.forEach(feedback -> feedback.setText(null));
+            }
+
+            // TODO: this is not good code!
+            setTestCaseCount(testCaseFeedback.size());
+            setPassedTestCaseCount((int) testCaseFeedback.stream().filter(feedback -> Boolean.TRUE.equals(feedback.isPositive())).count());
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/domain/Result.java:502]: Could be moved to a private helper method `filterTestCaseFeedback(ProgrammingExercise)`, but also okay as-is.

**Review by b-fein — APPROVED** [review]
Code changes look good.

**Review by laadvo — COMMENTED** [review]
Tested it in exam mode using different exercises, the test names are only shown in case the option to show the names was selected. Tested on ts1. 

I noticed that sometimes the task names disappear from the problem statement . I'm not sure if this PR has anything to do with it but thought I would mention it since I never experienced it before on production. 
image

**Comment by Strohgelaender** [comment]
Great finding @laadvo. This component needs the test case names in order to correctly create the mapping between tasks and testcases. I reverted the server changes and need to find a different solution for this.

**Review by simon-hng — APPROVED** [review]
Reapprove code

**Review by JohannesStoehr — APPROVED** [review]
Looked through the code, looks good to me!

**Review by DominikRemo — APPROVED** [review]
Tested in testing session on ts1. Worked as expected.

**Review by danielbode — APPROVED** [review]
Tested on ts1 during the testing session. Works as expected.

**Pull request merged**